### PR TITLE
make micro mode window load remote video

### PR DESCRIPTION
**Pull request by han-gyeol**
WIP. Don't merge!

### CODE REVIEW

File: config.js [hunk 1]
```diff
@@ -2,7 +2,7 @@ module.exports = {
     /**
      * The URL of the Jitsi Meet deployment that will be used.
      */
-    jitsiMeetURL: "https://meet.jit.si/",
+    jitsiMeetURL: "https://meet.jit.si/testurl",
```
Comment (bbaldino) [inline config.js]: probably didn't mean to commit this?

File: main.js [hunk 2]
```diff
@@ -35,6 +46,16 @@ const jitsiMeetWindowOptions = {
 };
 
 /**
+ * Options used when creating the micro mode window.
+ */
+const microWindowOptions = {
```
Comment (bbaldino) [inline main.js]: shouldn't there be something here about it being always on top?

File: main.js [hunk 3]
```diff
@@ -66,17 +87,55 @@ function setAPPListeners () {
  * Opens new window with index.html(Jitsi Meet is loaded in iframe there).
  */
 function createJitsiMeetWindow () {
-    jitsiMeetWindow = new BrowserWindow(jitsiMeetWindowOptions);
-    jitsiMeetWindow.loadURL(indexURL);
+  jitsiMeetWindow = new BrowserWindow(jitsiMeetWindowOptions);
```
Comment (bbaldino) [inline main.js]: it looks like all this got changed from 4 space indent to 2 space indent

File: main.js [hunk 4]
```diff
@@ -66,17 +87,55 @@ function setAPPListeners () {
...
+  jitsiMeetWindow.loadURL(indexURL);
+  microWindow = new BrowserWindow(microWindowOptions);
```
Comment (bbaldino) [inline main.js]: ok, so is this just always loading the micro window for now for testing? (as opposed to only loading it when the main window loses focus?)

File: main.js [hunk 5]
```diff
@@ -66,17 +87,55 @@ function setAPPListeners () {
...
+}
+
+/**
+ * Sets the ipc listeners for messages from renderer processes
+ */
+function setIPCListeners() {
+  // ipcMain.on('log', (event, data) => {
+  //   console.log(data);
+  // });
+  ipcMain.on('mainWindowPeerCreated', () => {
```
Comment (bbaldino) [inline main.js]: i wonder if there might be a better format for these messages where the 'destination' (main window vs micro window) was contained in a field rather than in the message name itself?

File: windows/jitsi-meet/index.html [hunk 6]
```diff
@@ -18,6 +18,7 @@
         </style>
     </head>
     <body>
+      <script src="https://webrtc.github.io/adapter/adapter-latest.js"></script>
```
Comment (bbaldino) [inline windows/jitsi-meet/index.html]: is this needed by the main window?  i would've thought we were already pulling in adapter in lib-jitsi-meet

File: windows/jitsi-meet/micro.js [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+
+/**
+ * The RTCPeerConnection to main window.
+ */
+let peerConnectionMicro;
+
+/**
+ * The remote video from main window.
+ */
+let largeVideo = document.createElement('video');
```
Comment (bbaldino) [inline windows/jitsi-meet/micro.js]: any reason not to just have the video element be defined in the html directly?

File: windows/jitsi-meet/micro.js [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+    function() {
+      ipcRenderer.send('log', "micro window setLocalDescription complete");
+    },
+    function(error) {
+      ipcRenderer.send('log', "micro window failed to set session description: " + error.toString());
+    }
+  );
+
+  const data = {type: desc.type, sdp: desc.sdp};
+  ipcRenderer.send('microWindowLocalDescriptionSet', data);
```
Comment (bbaldino) [inline windows/jitsi-meet/micro.js]: maybe refactor these calls out into a method like:

  function sendToMainWindow(messageName, data) {
    ipcRenderer.send(messageName, data);
  }

Comment (bbaldino) [inline windows/jitsi-meet/micro.js]: one for the logging calls might be good too?

**Comment by bbaldino** [comment]
made a few comments with some thoughts, probably mostly things you were planning on cleaning up in your refactor anyway.  looking good!

**Comment by hristoterezov** [comment]
@bbaldino Thanks for the comments! I asked @leook0209 to create the PR only for testing purposes. This is not the final version and he is working on fixing it. 


@leook0209 I'll add in the description that this is work in progress. This way everybody will know that this PR is not ready for merge/review.

File: windows/jitsi-meet/index.html (hunk 6 above)
Comment (leook0209) [inline windows/jitsi-meet/index.html]: As far as I tried, RTCPeerConnection was undefined without the adapter. I guess lib-jitsi-meet's adapter is contained within the iframe, so the Electron browser does not have access to it.

File: main.js (hunk 4 above)
Comment (leook0209) [inline main.js]: Currently my plan is to load the micro window when the main window loads, but hidden at first. When the user minimizes /lose focus the main window, I was going to make it visible. For the performance sake, I might gonna make an option to disable micro mode, which prevents the micro window from loading.

File: main.js (hunk 3 above)
Comment (leook0209) [inline main.js]: My mistake, will be using 4 space indent from now on

File: windows/jitsi-meet/micro.js (hunk 7 above)
Comment (leook0209) [inline windows/jitsi-meet/micro.js]: Nope, it is for convenience sake for development for now. In the end, all the element will be defined in html.

File: main.js [hunk 9]
```diff
@@ -31,7 +42,19 @@ const jitsiMeetWindowOptions = {
...
-    titleBarStyle: 'hidden'
+    titleBarStyle: 'hidden',
+    webPreferences: { experimentalFeatures: true }
+};
+
+/**
+ * Options used when creating the micro mode window.
+ */
+const microWindowOptions = {
+    width: 800,
```
Comment (saghul) [inline main.js]: we probably want it to be smaller, say 320 x 240. Not resizable.

File: main.js [hunk 10]
```diff
@@ -31,7 +42,19 @@ const jitsiMeetWindowOptions = {
...
+    height: 600,
+    titleBarStyle: 'hidden',
+    frame: true,
```
Comment (saghul) [inline main.js]: I think we should make it frameless. And make the entire window a drag area.

File: main.js [hunk 11]
```diff
@@ -31,7 +42,19 @@ const jitsiMeetWindowOptions = {
...
+    webPreferences: { experimentalFeatures: true }
```
Comment (saghul) [inline main.js]: we don't need this flag for the micro window itself, do we? The capture is done in the main window.

File: main.js (hunk 4 above)
Comment (saghul) [inline main.js]: I'd like us to test creating the window lazily. Having the window always there will make it "visible" to some window managers on Linux, for example.

File: main.js [hunk 12]
```diff
@@ -66,16 +89,48 @@ function setAPPListeners () {
...
+        if (data) {
+            microWindow.webContents.send(message, data);
+        } else {
+            microWindow.webContents.send(message);
+        }
+    });
+    ipcMain.on('microWindowEvent', (event, args) => {
+        const message = args[0];
+        const data = args[1];
+        if (data) {
```
Comment (saghul) [inline main.js]: why is this needed? doesn't sending undefined work?

File: modules/micromodeconnection/index.js [hunk 13]
```diff
@@ -0,0 +1,260 @@
...
+}
+
+/**
+ * Creates peer connection offer to micro window.
+ */
+function createAndSendOffer() {
+    log("main window createOffer start");
+
+    const offerOptions = {
+        offerToReceiveAudio: 1,
```
Comment (saghul) [inline modules/micromodeconnection/index.js]: we don't want to receive audio on this one, or the user will hear it twice.

File: modules/micromodeconnection/index.js [hunk 14]
```diff
@@ -0,0 +1,260 @@
...
+ */
+function sendMessageToMicroWindow(message, data) {
+  args = [];
+  args[0] = message;
+  args[1] = data;
+    ipcRenderer.send('mainWindowEvent', args);
+}
+
+
+/*------------------------------Micro Window----------------------------------*/
```
Comment (saghul) [inline modules/micromodeconnection/index.js]: can we split this file in 2? those globals are not nice to look at :-)

File: modules/micromodeconnection/index.js [hunk 15]
```diff
@@ -0,0 +1,260 @@
...
+    });
+
+    peerConnectionMicro.onicecandidate = function(event) {
+        if(event.candidate !== null) {
+            const newIceCandidate = {
+                candidate: event.candidate.candidate,
+                sdpMLineIndex: event.candidate.sdpMLineIndex,
+                sdpMid: event.candidate.sdpMid
+            };
+            sendMessageToMainWindow('microWindowIceCandidate', newIceCandidate);
```
Comment (saghul) [inline modules/micromodeconnection/index.js]: why not send the candidate just as it comes? it would also help with signaling the end of candidates.

Comment (saghul) [inline modules/micromodeconnection/index.js]: Also, we only want to send host candidates.

File: windows/jitsi-meet/index.html (hunk 6 above)
Comment (saghul) [inline windows/jitsi-meet/index.html]: We cannot do this. Things may break in different ways. If the Chrome version in Electron doesn't have RTCPeerConnection unprefixed, use webkitRTCPeerConnection please.

File: windows/jitsi-meet/micro.html [hunk 16]
```diff
@@ -0,0 +1,26 @@
...
+        <style>
+            body, html {
+                margin: 0;
+                height: 100%;
+                overflow: hidden;
+            }
+            body {
+                -webkit-app-region: drag
+            }
+            iframe {
```
Comment (saghul) [inline windows/jitsi-meet/micro.html]: why do we need an iframe here?

File: windows/jitsi-meet/micro.js (hunk 7 above)
Comment (saghul) [inline windows/jitsi-meet/micro.js]: I think you can move this to the HTML now, just query the selector in JS and do the needful there.

File: windows/jitsi-meet/render.js [hunk 17]
```diff
@@ -71,7 +72,16 @@ const dialogFactory = new DialogFactory();
  * Initializes postis.
  * Initializes remote control.
  */
+let largeVideo;
 function onload() {
+    largeVideo = iframe.contentWindow.document.getElementById("largeVideo");
+    // iframe.setAttribute('style', 'width:50%; height:50%; border:0; border:none');
+
+    const canvas = copyVideo(largeVideo, 400, 300);
```
Comment (saghul) [inline windows/jitsi-meet/render.js]: have you tried to do this without the canvas?

File: windows/jitsi-meet/render.js [hunk 18]
```diff
@@ -85,6 +95,24 @@ function onload() {
 }
 
 /**
+ * Create a copy of HTMLvideo in HTMLcanvas form
+ * Get reference of the video, width and height as parameters
+ * Draw canvas every 33ms
```
Comment (saghul) [inline windows/jitsi-meet/render.js]: Make this dynamic, so we can change it, we might want to do less than 30 fps.

File: .eslintrc.js [hunk 19]
```diff
@@ -19,19 +19,20 @@ module.exports = {
...
         // While it is considered a best practice to avoid using methods on
         // console in JavaScript that is designed to be executed in the browser
         // and ESLint includes the rule among its set of recommended rules, (1)
         // the general practice is to strip such calls before pushing to
         // production and (2) we prefer to utilize console in lib-jitsi-meet
         // (and jitsi-meet).
         'no-console': 'off',
-        'semi': 'error'
+        'semi': 'error',
+        'no-undef' : 'off'
```
Comment (leook0209) [inline .eslintrc.js]: Temporarily disabled some lint options because they conflicts with the webkitRTCPeerConnection class.

/home/han/git/jitsi-meet-electron/modules/micromodeconnection/index.js
  49:31  error  A constructor name should not start with a lowercase letter  new-cap
  49:31  error  'webkitRTCPeerConnection' is not defined                     no-undef

Not sure if I should remove them, or find a way around...?

File: main.js (hunk 4 above)
Comment (leook0209) [inline main.js]: What does it mean by 'lazily'? Is it creating the window when needed instead of at the beginning?

File: .eslintrc.js (hunk 19 above)
Comment (bbaldino) [inline .eslintrc.js]: typically there are ways to temporarily disable eslint rules in the files themselves (for exceptional cases), maybe that could be done instead? e.g. https://evanhahn.com/disable-eslint-for-a-file/

File: main.js [hunk 20]
```diff
@@ -31,7 +41,21 @@ const jitsiMeetWindowOptions = {
...
+/**
+ * Options used when creating the micro mode window.
+ */
+const microWindowOptions = {
+    width: 320,
+    height: 240,
+    titleBarStyle: 'hidden',
+    frame: false,
+    alwaysOnTop: true,
+    x: 1570,
```
Comment (bbaldino) [inline main.js]: are these absolute values for the position of the micro window?  wonder if we might get into trouble using absolute values here

File: main.js (hunk 4 above)
Comment (bbaldino) [inline main.js]: yup, pretty sure that's what saul means...it would definitely be interesting to try that and see how it performs.  if it doesn't perform all that well, maybe we could try some other options like creating the peer connection but only encoding/sending the media when the main window loses focus (to avoid the encoding cost all the time)

Comment (saghul) [inline main.js]: Yep, @bbaldino can read my mind :-)

File: package.json [hunk 21]
```diff
@@ -5,7 +5,7 @@
   "main": "main.js",
   "productName": "Jitsi Meet",
   "scripts": {
-    "start": "electron .",
+    "start": "electron --enable-transparent-visuals --disable-gpu .",
```
Comment (leook0209) [inline package.json]: Temporarily added a gpu disable command for linux (I use Ubuntu 16.04). Not tested yet for other OS

File: package.json [hunk 22]
```diff
@@ -34,9 +34,9 @@
     "postis": "^2.2.0"
   },
   "devDependencies": {
-    "electron": "1.4.13",
+    "electron": "*",
```
Comment (leook0209) [inline package.json]: Installing the latest version of electron doesn't really solve the video.captureStream() problem

File: windows/jitsi-meet/render.js [hunk 23]
```diff
@@ -71,7 +74,21 @@ const dialogFactory = new DialogFactory();
  * Initializes postis.
  * Initializes remote control.
  */
+let largeVideo;
+let mainWindow;
 function onload() {
+    largeVideo = iframe.contentWindow.document.getElementById("largeVideo");
+
+    ipcRenderer.on('hide', () => {
```
Comment (leook0209) [inline windows/jitsi-meet/render.js]: Lazily create a p2p connection to Micro window

File: windows/jitsi-meet/render.js [hunk 24]
```diff
@@ -71,7 +74,21 @@ const dialogFactory = new DialogFactory();
...
+        setupMicroModePeerConnection(largeVideo.srcObject);
+    });
+
+    largeVideo.ondurationchange = function() {
```
Comment (leook0209) [inline windows/jitsi-meet/render.js]: this has solved the video transition problem

File: windows/jitsi-meet/render.js [hunk 25]
```diff
@@ -84,6 +101,34 @@ function onload() {
         config.handleRemoteControlAuthorization);
 }
 
+function setupMicroModePeerConnection (stream) {
+    mainWindow = new WindowPeerConnection('jitsiMeetWindow');
+    mainWindow.attachStream(stream);
+    mainWindow.sendStream('microWindow');
+}
+
+function switchMicroVideo (stream) {
```
Comment (leook0209) [inline windows/jitsi-meet/render.js]: Resend a new MediaStream to Micro window

File: windows/jitsi-meet/render.js [hunk 26]
```diff
@@ -1,23 +1,32 @@
...
-let iframe = document.createElement('iframe');
-iframe.src = process.env.JITSI_MEET_URL || config.jitsiMeetURL;
-iframe.allowFullscreen = true;
-iframe.onload = onload;
-document.body.appendChild(iframe);
+ var options = {
+     domain: config.jitsiMeetURL,
+     room: "testurl",
+     width: '100%',
+     height: '100%'
```
Comment (leook0209) [inline windows/jitsi-meet/render.js]: It seems like width and height options doesn't work for jitsi-meet ExternalAPI. Regardless of which number I give, it always creates an iframe of MIN width and height, 790 * 300.

File: windows/jitsi-meet/render.js [hunk 27]
```diff
@@ -17,13 +17,15 @@ let channel;
...
-};
-let JitsiMeetAPI = new JitsiMeetExternalAPI(config.jitsiMeetURL, options);
+let domain = config.jitsiMeetURL;
+let room = "testurl";
+let width = '100%';
+let height = '100%';
+let htmlElement = undefined;
+let configOverwrite = {};
+let interfaceConfigOverwrite = {};
+let JitsiMeetAPI = new JitsiMeetExternalAPI(domain, room, width, height,
```
Comment (hristoterezov) [inline windows/jitsi-meet/render.js]: Why do you need this change?

Comment (leook0209) [inline windows/jitsi-meet/render.js]: Using options object somehow didnt really work out, some of the parameters were ignored in the iframe created.

Comment (leook0209) [inline windows/jitsi-meet/render.js]: Oh, if you were asking about using the JitsiMeetExternalAPI, it was necessary for Micro window's toolbar buttons to manipulate the Jtisi-meet application
